task/thermal/src/control.rs

```diff
@@ -116,84 +142,216 @@ impl InputChannel {
-///         |
-///         0------ Component margin + target_margin --
-///         |
-///         |   Fan speed increases slowly
-///         |
-///         ------- slow_band -------------------------
-///         |
-///         |   Fan speed remains constant
-///         |
-///         ------- dead_band -------------------------
-///         |
-///   COLD  |   Fan speed decreases slowly
-///         |
-/// ```
-///
-pub(crate) struct ThermalControl<'a, B> {
+pub(crate) struct ThermalControl<'a> {
     /// Reference to board-specific parameters
-    bsp: &'a B,
+    bsp: &'a Bsp,
 
     /// Task to which we should post sensor data updates
     sensor_api: SensorApi,
 
-    /// Dead band between increasing and decreasing fan speed
-    dead_band: Celsius,
-
-    /// Band where fan speed increases slowly
-    slow_band: Celsius,
-
     /// Target temperature margin. This must be >= 0; as it increases, parts
-    /// are kept cooler than their max temperature ratings.
+    /// are kept cooler than their target temperature value.
     target_margin: Celsius,
 
-    /// Commanded PWM value (0-100) for every output channel
-    target_pwm: u8,
+    /// Controller state
+    state: ThermalControlState,
+
+    /// How long to wait in the `Overheated` state before powering down
+    overheat_timeout_ms: u64,
 
+    /// Once we're in `Overheated`, how much does the temperature have to drop
+    /// by before we return to `Normal`
+    overheat_hysteresis: Celsius,
+
+    /// Most recent power mode mask
+    power_mode: u32,
+
+    /// Number of sensor reads which failed
     read_failed_count: u32,
+
+    /// Number of sensor posts which failed
     post_failed_count: u32,
 }
 
-impl<'a, B: BspT> ThermalControl<'a, B> {
+/// Represents a temperature reading at the time at which it was taken
+#[derive(Copy, Clone, Debug)]
+enum TemperatureReading {
+    /// Normal reading, timestamped using monotonic system time
+    Valid { time_ms: u64, value: Celsius },
+
+    /// This sensor is not used in the current power state
+    Inactive,
+}
+
+/// Configuration for a PID controller
+pub struct PidConfig {
+    pub gain_p: f32,
+    pub gain_i: f32,
+    pub gain_d: f32,
+}
+
+struct PidState {
+    /// Previous (time, input) tuple, for derivative term
+    prev_error: Option<f32>,
+
+    /// Accumulated integral term
+    integral: f32,
+}
+
+impl PidState {
+    /// Attempts to drive the error to zero.
+    ///
+    /// The error and output are expected to have the same signs, i.e. a large
+    /// positive error will produce a large positive output.
+    fn run(
+        &mut self,
+        cfg: &PidConfig,
+        error: f32,
+        limit: core::ops::Range<f32>,
+    ) -> f32 {
+        let p_contribution = cfg.gain_p * error;
+
+        // Only accumulate the integral term if we're not saturated, to prevent
+        // integral wind-up.
+        if limit.contains(&p_contribution) {
+            self.integral += error;
```

Comment:
Guarding against overflow/underflow? This is a float, so maybe we're always safe.

Reply:
I've dramatically reworked this section, so it now clamps in a different way.